crates/build-rs/src/input.rs

```diff
@@ -480,112 +445,159 @@ pub fn rustc_wrapper() -> Option<PathBuf> {
 #[track_caller]
 pub fn cargo_pkg_version_patch() -> u64 {
-    get_num("CARGO_PKG_VERSION_PATCH")
+    to_parsed(var_or_panic("CARGO_PKG_VERSION_PATCH"))
 }
 
 /// The pre-release version of your package.
 #[track_caller]
-pub fn cargo_pkg_version_pre() -> String {
-    get_str("CARGO_PKG_VERSION_PRE")
+pub fn cargo_pkg_version_pre() -> Option<String> {
+    to_opt(var_or_panic("CARGO_PKG_VERSION_PRE")).map(to_string)
 }
 
 /// Colon separated list of authors from the manifest of your package.
 #[track_caller]
 pub fn cargo_pkg_authors() -> Vec<String> {
-    get_str("CARGO_PKG_AUTHORS")
-        .split(':')
-        .map(str::to_owned)
-        .collect()
+    to_strings(var_or_panic("CARGO_PKG_AUTHORS"), ':')
 }
 
 /// The name of your package.
 #[track_caller]
 pub fn cargo_pkg_name() -> String {
-    get_str("CARGO_PKG_NAME")
+    to_string(var_or_panic("CARGO_PKG_NAME"))
 }
 
 /// The description from the manifest of your package.
 #[track_caller]
-pub fn cargo_pkg_description() -> String {
-    get_str("CARGO_PKG_DESCRIPTION")
+pub fn cargo_pkg_description() -> Option<String> {
+    to_opt(var_or_panic("CARGO_PKG_DESCRIPTION")).map(to_string)
 }
 
 /// The home page from the manifest of your package.
 #[track_caller]
-pub fn cargo_pkg_homepage() -> String {
-    get_str("CARGO_PKG_HOMEPAGE")
+pub fn cargo_pkg_homepage() -> Option<String> {
+    to_opt(var_or_panic("CARGO_PKG_HOMEPAGE")).map(to_string)
 }
 
 /// The repository from the manifest of your package.
 #[track_caller]
-pub fn cargo_pkg_repository() -> String {
-    get_str("CARGO_PKG_REPOSITORY")
+pub fn cargo_pkg_repository() -> Option<String> {
+    to_opt(var_or_panic("CARGO_PKG_REPOSITORY")).map(to_string)
 }
 
 /// The license from the manifest of your package.
 #[track_caller]
-pub fn cargo_pkg_license() -> String {
-    get_str("CARGO_PKG_LICENSE")
+pub fn cargo_pkg_license() -> Option<String> {
+    to_opt(var_or_panic("CARGO_PKG_LICENSE")).map(to_string)
 }
 
 /// The license file from the manifest of your package.
 #[track_caller]
-pub fn cargo_pkg_license_file() -> PathBuf {
-    get_path("CARGO_PKG_LICENSE_FILE")
+pub fn cargo_pkg_license_file() -> Option<PathBuf> {
+    to_opt(var_or_panic("CARGO_PKG_LICENSE_FILE")).map(to_path)
 }
 
 /// The Rust version from the manifest of your package. Note that this is the
 /// minimum Rust version supported by the package, not the current Rust version.
 #[track_caller]
-pub fn cargo_pkg_rust_version() -> String {
-    get_str("CARGO_PKG_RUST_VERSION")
+pub fn cargo_pkg_rust_version() -> Option<String> {
+    to_opt(var_or_panic("CARGO_PKG_RUST_VERSION")).map(to_string)
 }
 
 /// Path to the README file of your package.
 #[track_caller]
-pub fn cargo_pkg_readme() -> PathBuf {
-    get_path("CARGO_PKG_README")
+pub fn cargo_pkg_readme() -> Option<PathBuf> {
+    to_opt(var_or_panic("CARGO_PKG_README")).map(to_path)
+}
+
+fn is_present(key: &str) -> bool {
+    var_os(key).is_some()
+}
+
+#[track_caller]
```

Comment:
Note to reviewers: buildscripts are compiled without debug info by default, so backtraces aren't helpful. That's why `#[track_caller]` is applied so liberally here, so the caller actually knows where the bad call was made.

Reply:
I wonder where we should record design principles like that.  For example, I had considered moving to `env!` but realized that re-running is cheaper than recompilation

Reply:
The default answer would be a DESIGN.md alongside README.md. That seems as reasonable a choice as any other[^1], although any such document always is difficult to organize and keep up to date.

[^1]: The main alternative being a wiki page, which is even less likely to be updated or even looked at ime since it isn't colocated.

As for `env!` — it's lots of "fun" to keep track of whether we're querying the environment for building build-rs (set at our compile time) or for the client buildscript (only set at our run time).

Reply:
For build-rs I think we should in general query only runtime envs, though I don't fully get the context of using `env!` for this function.

Reply:
The only time we use `env!` is to get `build-rs`s MSRV.  If we were to do things at compile time, we'd have to turn all of these functions into macros so that `env!` runs in the callers context.